src/tools/wasm-opt.cpp

```diff
@@ -117,6 +118,15 @@ int main(int argc, const char* argv[]) {
          WasmOptOption,
          Options::Arguments::Zero,
          [&](Options* o, const std::string& arguments) { converge = true; })
+    .add("--preserve-type-order",
+         "",
+         "Preserve the order of types from the input (useful for debugging and "
+         "testing)",
+         WasmOptOption,
+         Options::Arguments::Zero,
+         [&](Options* o, const std::string& arguments) {
+           preserveTypeOrder = true;
+         })
```

Comment:
This could perhaps be in `tool-options.h` as it can be useful in any commandline tool that writes a binary, like `wasm-merge`, `wasm-split`, etc.?

Reply:
Perhaps, although I thought it would be simpler to just have those tools unconditionally maintain the type order.

Reply:
Ah, sounds good, that's what I was missing. Let me re-read it.

Reply:
Though, perhaps flip the option then? If all other tools preserve by default, maybe wasm-opt should do the same, and have a flag to *not* preserve the order.

Reply:
Maybe, although as a matter of practicality I think it's better the way it is. If we flipped the default, the diff in this PR would get much larger. Also, users would never want the order preserved in their production builds. If we changed the default to preserve type order, everyone's builds would regress until they added the new flag. Unless we had the optimization flags like `-O2` imply that the type order should not be preserved, but that's getting more complicated to remember and would inhibit our ability to use the flag to improve test readability.

Reply:
> Also, users would never want the order preserved in their production builds.

Ah, that's a critical point that I didn't appreciate before. In fact, it's so crucial that I think we should change the other tools, since people can emit things from `wasm-split/merge/etc.` and send them straight to production?

Reply:
Ok, in that case I will add this to tool-options.h.